public/app/features/trails/Integrations/logsIntegration.ts

```diff
@@ -0,0 +1,139 @@
+import { lastValueFrom } from 'rxjs';
+
+import { DataSourceSettings } from '@grafana/data';
+import { BackendSrvRequest, FetchResponse, getBackendSrv } from '@grafana/runtime';
+
+export type RecordingRuleGroup = {
+  name: string;
+  rules: RecordingRule[];
+};
+
+export type RecordingRule = {
+  name: string;
+  query: string;
+  type: string; // can be alerting or recording
+};
+
+export type FoundLokiDataSource = Pick<DataSourceSettings, 'name' | 'uid'>;
+export type ExtractedRecordingRule = RecordingRule & { datasource: FoundLokiDataSource };
+export type ExtractedRecordingRules = {
+  [dataSourceID: string]: ExtractedRecordingRule[];
+};
+
+export async function fetchLokiDataSources(): Promise<DataSourceSettings[]> {
+  const dataSourcesReq: BackendSrvRequest = { url: 'api/datasources' };
+  const { data } = await lastValueFrom<FetchResponse<DataSourceSettings[]>>(getBackendSrv().fetch(dataSourcesReq));
+  return data.filter((d) => d.type === 'loki');
+}
+
+export function buildRecordingRuleURL(datasourceSettings: DataSourceSettings): string {
+  return `api/prometheus/${datasourceSettings.uid}/api/v1/rules`;
+}
+
+export async function fetchLokiRecordingRules(url: string) {
+  const recordingRules: BackendSrvRequest = { url };
+  const { data } = await lastValueFrom<
+    FetchResponse<{
+      data: { groups: RecordingRuleGroup[] };
+    }>
+  >(getBackendSrv().fetch(recordingRules));
+  return data.data.groups;
+}
+
+export function extractRecordingRules(
+  ruleGroups: RecordingRuleGroup[],
+  ds: DataSourceSettings
+): ExtractedRecordingRule[] {
+  if (ruleGroups.length === 0) {
+    return [];
+  }
+
+  const extractedRules: ExtractedRecordingRule[] = [];
+  ruleGroups.forEach((rg) => {
+    rg.rules
+      .filter((r) => r.type === 'recording')
+      .forEach(({ type, name, query }) => {
+        extractedRules.push({
+          type,
+          name,
+          query,
+          datasource: {
+            name: ds.name,
+            uid: ds.uid,
+          },
+        });
+      });
+  });
+
+  return extractedRules;
+}
+
+function extractLokiQueryFromRecordingRule(rule: ExtractedRecordingRule): string {
+  // Remove Unicode escapes and unneeded quotes
+  let cleanedRule = rule.query.replace(/\\u[\dA-Fa-f]{4}/g, (match) => {
+    return String.fromCharCode(parseInt(match.replace('\\u', ''), 16));
+  });
+  cleanedRule = cleanedRule
+    .replace(/\\"/g, '"') // Replace escaped quotes
+    .replace(/^"|"$/g, ''); // Remove starting and ending quotes
+
+  // Extract the Loki query
+  const queryRegex = /{[^}]+}.*?(?=\[|$)/;
+  const match = cleanedRule.match(queryRegex);
+  const extractedQuery = match ? match[0].trim() : '';
+
+  return extractedQuery;
+}
```

Comment:
We do have some good helper functions for that in e.g. https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/queryUtils.ts#L247-L263

Reply:
Thanks for sharing this - I was hoping that these existed. When the time comes to [externalize](https://docs.google.com/document/d/1M0i2XJ-lseb8gtCqk02vxxNUsWk5P7dbUP8449qJ9Rw/edit?usp=sharing) Explore Metrics, we will need to copy them. If these utils are ever extracted into a package, that would be ideal.

Reply:
Implemented in https://github.com/grafana/grafana/pull/94656/commits/1a08ae5373be7bc26ed3b5d70ee4876b167d4806